### ISSUE TITLE
Lancero driver won't build on Linux 4.13

### ISSUE THREAD

**Issue by joefowler** [usnistgov/quantum-sensors-drivers#7]
**[Original report](https://bitbucket.org/nist_microcal/nist_qsp_drivers/issues/7) by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

The Linux kernel API has changed in some way. Cannot build lancero driver on kernel 4.13.0-41. @darktyr might want to follow this issue.

```
pcuser@687raven:~/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver$ make
make -C /lib/modules/4.13.0-41-generic/build M=/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver modules
make[1]: Entering directory '/usr/src/linux-headers-4.13.0-41-generic'
  CC [M]  /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.o
In file included from ./include/linux/mutex.h:13:0,
                 from ./include/linux/kernfs.h:13,
                 from ./include/linux/sysfs.h:15,
                 from ./include/linux/kobject.h:21,
                 from ./include/linux/cdev.h:4,
                 from /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:12:
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c: In function ‘sgm_get_user_pages’:
./arch/x86/include/asm/current.h:17:17: warning: passing argument 1 of ‘get_user_pages’ makes integer from pointer without a cast [-Wint-conversion]
 #define current get_current()
                 ^
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:145:22: note: in expansion of macro ‘current’
  rc = get_user_pages(current, current->mm, (unsigned long)start, nr_pages, to_u
                      ^
In file included from ./include/linux/scatterlist.h:7:0,
                 from ./include/linux/dma-mapping.h:10,
                 from /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:13:
./include/linux/mm.h:1343:6: note: expected ‘long unsigned int’ but argument is of type ‘struct task_struct *’
 long get_user_pages(unsigned long start, unsigned long nr_pages,
      ^
In file included from ./include/linux/mutex.h:13:0,
                 from ./include/linux/kernfs.h:13,
                 from ./include/linux/sysfs.h:15,
                 from ./include/linux/kobject.h:21,
                 from ./include/linux/cdev.h:4,
                 from /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:12:
./arch/x86/include/asm/current.h:17:17: warning: passing argument 2 of ‘get_user_pages’ makes integer from pointer without a cast [-Wint-conversion]
 #define current get_current()
                 ^
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:145:31: note: in expansion of macro ‘current’
  rc = get_user_pages(current, current->mm, (unsigned long)start, nr_pages, to_u
                               ^
In file included from ./include/linux/scatterlist.h:7:0,
                 from ./include/linux/dma-mapping.h:10,
                 from /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:13:
./include/linux/mm.h:1343:6: note: expected ‘long unsigned int’ but argument is of type ‘struct mm_struct *’
 long get_user_pages(unsigned long start, unsigned long nr_pages,
      ^
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:145:66: warning: passing argument 4 of ‘get_user_pages’ makes pointer from integer without a cast [-Wint-conversion]
  rc = get_user_pages(current, current->mm, (unsigned long)start, nr_pages, to_u
                                                                  ^
In file included from ./include/linux/scatterlist.h:7:0,
                 from ./include/linux/dma-mapping.h:10,
                 from /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:13:
./include/linux/mm.h:1343:6: note: expected ‘struct page **’ but argument is of type ‘int’
 long get_user_pages(unsigned long start, unsigned long nr_pages,
      ^
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:145:76: warning: passing argument 5 of ‘get_user_pages’ makes pointer from integer without a cast [-Wint-conversion]
 = get_user_pages(current, current->mm, (unsigned long)start, nr_pages, to_user,
                                                                        ^
In file included from ./include/linux/scatterlist.h:7:0,
                 from ./include/linux/dma-mapping.h:10,
                 from /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:13:
./include/linux/mm.h:1343:6: note: expected ‘struct vm_area_struct **’ but argument is of type ‘int’
 long get_user_pages(unsigned long start, unsigned long nr_pages,
      ^
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:145:7: error: too many arguments to function ‘get_user_pages’
  rc = get_user_pages(current, current->mm, (unsigned long)start, nr_pages, to_u
       ^
In file included from ./include/linux/scatterlist.h:7:0,
                 from ./include/linux/dma-mapping.h:10,
                 from /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:13:
./include/linux/mm.h:1343:6: note: declared here
 long get_user_pages(unsigned long start, unsigned long nr_pages,
      ^
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c: In function ‘sgm_kernel_pages’:
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:249:24: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  unsigned char *virt = start;
                        ^
scripts/Makefile.build:316: recipe for target '/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.o' failed
make[2]: *** [/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.o] Error 1
Makefile:1550: recipe for target '_module_/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver' failed
make[1]: *** [_module_/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-4.13.0-41-generic'
Makefile:32: recipe for target 'lancero.ko' failed
make: *** [lancero.ko] Error 2
```

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

See [this commit](https://github.com/torvalds/linux/commit/d4edcf0d56958db0aca0196314ca38a5e730ea92) to the Linux kernel, during the 4.6 development.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Partly fixed Lancero for kernel 4.9+ and 4.13+. Addresses #7.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Fix some kernel driver compile warnings.

Fixes #7. Also replace erroneous spaces with TAB in Makefile.
